src/Topologies/Topologies.jl

```diff
@@ -166,10 +166,340 @@ struct Vertex{T <: AbstractTopology, V}
     num::V
 end
 
+"""
+    TensorProductTopology(mesh)
+
+A generic tensor-product topology defined on either an equispaced, regular rectangular
+structured grid or irregular one.
+"""
+struct TensorProductTopology{M <: AbstractMesh, B} <: AbstractTopology
```

Comment:
again, can we move tensor-product specific code to a new file?

Reply:
In commit [7252328](https://github.com/CliMA/ClimaCore.jl/pull/82/commits/7252328b79e5ad50d5615fe53a03241b09bc0492), I renamed `GridTopology` -> `TensorProductTopology` since their constructors and behaviors were the same. I moved the common methods to this file, and separated only the methods that needed to be dispatched based on the distinct, concrete types in the corresponding files `grid.jl` and `tensorproductmesh.jl`. Where would the common code reside then?

Reply:
Like the common methods in a separate common file, say `Topologies/grids.jl` and the specific methods (basically only `vertex_coordinates` and `opposing_face` are left to be distinct)  in `Topologies/regulargrid.jl` (what is now `Topologies/grid.jl`) and `Topologies/tensorproduct.jl`?

Reply:
Okay, I've done this in [`945407c`](https://github.com/CliMA/ClimaCore.jl/pull/82/commits/945407c14707b0234f7887b90c68dbc4643db408). I hope it is clearer now.

Reply:
In [`82f89ed`](https://github.com/CliMA/ClimaCore.jl/pull/82/commits/82f89ed04ce33546240e0489473b174aad391484), I put both `GridTopology` implementations (ie, the one dispatching on the `EquispacedRectangularMesh` and the one dispatching on the `TensorProductMesh` back into the same `Topologies/grids.jl` file. I hope it is clearer now.